File: No_2976/Minimum Cost to Convert String I.cpp
class Solution {
public:
    long long minimumCost(string source, string target, vector<char>& original, vector<char>& changed, vector<int>& cost) {
        int numOfVertex = 26;
        vector<vector<int>> dis(numOfVertex, vector<int>(numOfVertex, INT_MAX));
        auto x = [] (char c) {
            return c - 'a';
        };
        
        for (int i = 0; i < original.size(); i++)
            if (dis[x(original[i])][x(changed[i])] > cost[i])
                dis[x(original[i])][x(changed[i])] = cost[i];
        
        for (int i = 0; i < numOfVertex; i++)
            dis[i][i] = 0;
        
        for (int k = 0; k < numOfVertex; k++) {
            for (int i = 0; i < numOfVertex; i++) {
                for (int j = 0; j < numOfVertex; j++) {
                     if(dis[i][k] != INT_MAX && dis[k][j] != INT_MAX)
                        dis[i][j] = min(dis[i][j], dis[i][k] + dis[k][j]);
                }
            }
        }
        
        long long ret = 0;
        for (int i = 0; i < source.length(); i++) {
            if (dis[x(source[i])][x(target[i])] == INT_MAX)
                return -1;
            ret += dis[x(source[i])][x(target[i])];
        }
        
        return ret;
    }
};

File: No_2976/Minimum Cost to Convert String I.py
class Solution:
    def minimumCost(self, source: str, target: str, original: List[str], changed: List[str], cost: List[int]) -> int:
        NUM_OF_VERTEX = 26
        length = len(original)
        dis = defaultdict(lambda: float('inf'))
        
        for i, j, w in zip(original, changed, cost):
            if dis[(i, j)] > w:
                dis[(i, j)] = w
            
        for v in set(source):
            dis[(v, v)] = 0
        
        for k in ascii_lowercase:
            for i in ascii_lowercase:
                for j in ascii_lowercase:
                    dis[(i, j)] = min(dis[(i, j)], dis[(i, k)] + dis[(k, j)])
        ret = 0
        for s, t in zip(source, target):
            if dis[(s, t)] == float('inf'):
                return -1
            ret += dis[(s, t)]
            
        return ret